// @ngx-config/core/esm5/config.service.js
import { Injectable } from '@angular/core';
import { ConfigLoader } from './config.loader';
import * as i0 from "@angular/core";
import * as i1 from "./config.loader";
var ConfigService = (function () {
    function ConfigService(loader) {
        this.loader = loader;
    }
    ConfigService.prototype.init = function () {
        var _this = this;
        return this.loader.loadSettings().then(function (res) { return (_this.settings = res); });
    };
    ConfigService.prototype.getSettings = function (key, defaultValue) {
        if (!key || (Array.isArray(key) && !key[0])) {
            return this.settings;
        }
        var paths = !Array.isArray(key) ? key.split('.') : key;
        var result = paths.reduce(function (acc, current) { return acc && acc[current]; }, this.settings);
        if (result === undefined) {
            result = defaultValue;
            if (result === undefined) {
                throw new Error("No setting found with the specified key [" + paths.join('/') + "]!");
            }
        }
        return result;
    };
    ConfigService.ɵfac = function ConfigService_Factory(t) { return new (t || ConfigService)(i0.ɵɵinject(i1.ConfigLoader)); };
    ConfigService.ɵprov = i0.ɵɵdefineInjectable({ token: ConfigService, factory: ConfigService.ɵfac });
    return ConfigService;
}());
export { ConfigService };
(function () { i0.ɵsetClassMetadata(ConfigService, [{
        type: Injectable
    }], function () { return [{ type: i1.ConfigLoader }]; }, null); })();
