// @ngx-config/core/esm2015/config.service.js
import { Injectable } from '@angular/core';
import { ConfigLoader } from './config.loader';
import * as i0 from "@angular/core";
import * as i1 from "./config.loader";
export class ConfigService {
    constructor(loader) {
        this.loader = loader;
    }
    init() {
        return this.loader.loadSettings().then((res) => (this.settings = res));
    }
    getSettings(key, defaultValue) {
        if (!key || (Array.isArray(key) && !key[0])) {
            return this.settings;
        }
        const paths = !Array.isArray(key) ? key.split('.') : key;
        let result = paths.reduce((acc, current) => acc && acc[current], this.settings);
        if (result === undefined) {
            result = defaultValue;
            if (result === undefined) {
                throw new Error(`No setting found with the specified key [${paths.join('/')}]!`);
            }
        }
        return result;
    }
}
ConfigService.ɵfac = function ConfigService_Factory(t) { return new (t || ConfigService)(i0.ɵɵinject(i1.ConfigLoader)); };
ConfigService.ɵprov = i0.ɵɵdefineInjectable({ token: ConfigService, factory: ConfigService.ɵfac });
(function () { i0.ɵsetClassMetadata(ConfigService, [{
        type: Injectable
    }], function () { return [{ type: i1.ConfigLoader }]; }, null); })();
