// @ngx-config/core/esm2015/config.module.js
import { APP_INITIALIZER, Inject, InjectionToken, NgModule, Optional, SkipSelf } from '@angular/core';
import { ConfigLoader, ConfigStaticLoader } from './config.loader';
import { ConfigPipe } from './config.pipe';
import { ConfigService } from './config.service';
import * as i0 from "@angular/core";
export const configFactory = () => new ConfigStaticLoader();
export const initializerFactory = (config) => {
    const res = () => config.init();
    return res;
};
export const CONFIG_FORROOT_GUARD = new InjectionToken('CONFIG_FORROOT_GUARD');
export function provideForRootGuard(config) {
    if (config) {
        throw new Error(`ConfigModule.forRoot() called twice. Lazy loaded modules should use ConfigModule.forChild() instead.`);
    }
    return 'guarded';
}
export class ConfigModule {
    constructor(guard) { }
    static forRoot(configuredProvider = {
        provide: ConfigLoader,
        useFactory: configFactory
    }) {
        return {
            ngModule: ConfigModule,
            providers: [
                configuredProvider,
                ConfigService,
                {
                    provide: APP_INITIALIZER,
                    useFactory: initializerFactory,
                    deps: [ConfigService],
                    multi: true
                },
                {
                    provide: CONFIG_FORROOT_GUARD,
                    useFactory: provideForRootGuard,
                    deps: [[ConfigService, new Optional(), new SkipSelf()]]
                }
            ]
        };
    }
    static forChild() {
        return {
            ngModule: ConfigModule
        };
    }
}
ConfigModule.ɵmod = i0.ɵɵdefineNgModule({ type: ConfigModule });
ConfigModule.ɵinj = i0.ɵɵdefineInjector({ factory: function ConfigModule_Factory(t) { return new (t || ConfigModule)(i0.ɵɵinject(CONFIG_FORROOT_GUARD, 8)); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && i0.ɵɵsetNgModuleScope(ConfigModule, { declarations: [ConfigPipe], exports: [ConfigPipe] }); })();
(function () { i0.ɵsetClassMetadata(ConfigModule, [{
        type: NgModule,
        args: [{
                declarations: [ConfigPipe],
                exports: [ConfigPipe]
            }]
    }], function () { return [{ type: undefined, decorators: [{
                type: Optional
            }, {
                type: Inject,
                args: [CONFIG_FORROOT_GUARD]
            }] }]; }, null); })();
